### ISSUE TITLE
Suggestion: "soft" highlight first implicit candidate when none are selected

### ISSUE THREAD

**Issue by jdtsmith** [minad/corfu#45]
Here's company's take (no selection):

![image](https://user-images.githubusercontent.com/93749/126868532-dc45a2e9-c665-4677-ada9-90ccdcd286ab.png)

I.e. the 1st candidate is _always_ selected no matter what.  That doesn't work for corfu, because of the auto-commit-if-candidate-selected feature, but another less prominent selection face could be used to convey "if you hit `[Ret]`, you will have the first match inserted" (but never auto-committed!).

**Comment by minad** [comment]
I tried this a few days ago and I didn't really like it since it adds zero value, only visual noise.

I also thought about another UI mode: Do not show the current selection in the input line and highlight the first candidate directly. But I think I like the current way more. Supporting different frontend display modes is something I would like to avoid for Corfu. I generally want Corfu to be configurable where it is cheap to have (corfu-quit-..., corfu-commit-predicate, etc), but I would like to avoid adding intrusive configuration options which affect many places and introduce many code paths.

**Comment by jdtsmith** [comment]
Perhaps I wasn't clear enough: what I envision would be a pure UI flourish solely within `corfu--popup-show` — if none are selected, highlight the first candidate faintly.  See #50.  

Here's how it looks (dark):

![image](https://user-images.githubusercontent.com/93749/128724442-f7c9d5c1-8df0-4104-9d83-aac9f0b1f4b9.png)

vs. actual selection:

![image](https://user-images.githubusercontent.com/93749/128724910-91202013-31f4-4975-85b1-8479fdf21ae3.png)

**Comment by minad** [comment]
No, I really don't like this. We should not add this. Maybe a better idea comes up regarding the implicit preselection of the first candidate. I also experimented with showing an overlay which extends the candidate "find-a|lias-regexp", where the "lias-regexp" is shown in faint color. However the issue is that this works only for prefix completion.

**Comment by jdtsmith** [comment]
Sure.  Up to you of course, but you might consider checking in with other users to see how much mental overhead having _nothing_ selected gives.  But maybe I've just been using company too long.

**Comment by minad** [comment]
I argue it gives not much mental overhead - the popup already is the indication. I don't want to close the door on other alternatives now and when I tried it, I didn't like the noise of the additional highlighting. The issue is, the highlighting adds zero information, it is only eye candy and takes the focus from the prompt.

**Comment by minad** [comment]
Now that I think about this again, I also found it odd that given the preselected candidate if you press up/down, the key selects this candidate and doesn't jump over it. All in all it didn't feel like a proper solution, there should be something better.

How does Company solve this? I haven't tried it for a while. And back then I used a heavily modified configuration. This was also part of the motivation to make Corfu, since the Company configuration was way too complex.

**Comment by jdtsmith** [comment]
One man's noise is another's mental hitch soother ;).  It's a reasonable point about moving selection but not moving. 

company doesn't have the notion of "nothing selected".  Something is _always_ selected; the first item by default. So down-arrow just moves to the 2nd item.  But company can get away with this, because it does not change the input text at prompt in the buffer when the popup is active, so there is no need to discriminate between "something selected" and "nothing selected".  Some might argue that is the simpler approach.  

The one time company _does_ put a selection into the buffer is when there comes to be one and only one candidate, but only if the user never made a selection with arrows/whatever before that.  It puts that candidate there with some underlining.  But you still have to hit Return to accept it.  

Complete agreement about corfu being simpler to configure/maintain, and keeping that way should be a strong principle.

**Comment by jdtsmith** [comment]
> Note that the prompt _is selected_.

I don't know what this means.  How is the prompt selected?  I see no selection (though implicitly `ffap-el` is). 

![image](https://user-images.githubusercontent.com/93749/128753090-441eaa4a-9027-4f10-a357-521e041e3299.png)
 
BTW, should probably also call/display the doc-string for the first candidate when nothing is selected.

**Comment by minad** [comment]
> The one time company does put a selection into the buffer is when there comes to be one and only one candidate, but only if the user never made a selection with arrows/whatever before that. It puts that candidate there with some underlining. But you still have to hit Return to accept it.

I've tried this. But didn't feel very useful given that powerful completion styles give many candidates and don't only return the prefix matches. 

I see that the current Corfu is a bit odd regarding selection of the first candidate. On the other hand, overwriting the prompt as Corfu does feels much better for auto commit and also for commit with RET, since you directly see what you will get. I believe Company also allows this as an option?

**Comment by minad** [comment]
> I don't know what this means. How is the prompt selected? I see no selection (though implicitly ffap-el is).

Internally the prompt is selected implicitly. If you type, you append to the prompt. The oddity is that RET is special cased, since RET on the prompt would do nothing, so in this case the first candidate is used, which led you to request the soft highlighting.

**Comment by minad** [comment]
> BTW, should probably also call/display the doc-string for the first candidate when nothing is selected.

I am not sure. As long as you move the point to the desired candidate, I don't see a reason to show the docstring. The docstring is also noise, it is good to not show that right away.

**Comment by jdtsmith** [comment]
> > The one time company does put a selection into the buffer is when there comes to be one and only one candidate, but only if the user never made a selection with arrows/whatever before that. It puts that candidate there with some underlining. But you still have to hit Return to accept it.
> 
> I've tried this. But didn't feel very useful given that powerful completion styles give many candidates and don't only return the prefix matches.

Agree.  

> I see that the current Corfu is a bit odd regarding selection of the first candidate. On the other hand, overwriting the prompt as Corfu does feels much better for auto commit and also for commit with RET, since you directly see what you will get. 

Well the point is that you don't "see what you will get" when you haven't yet selected anything!  I tend to use orderless narrowing until my match is at the top, then hit Return.  In that case, I've never seen any indication that the top item is selected, neither by highlighting in the popup, nor in-buffer prompt text.  I can hit tab of course (but then I still need to hit Return). 

> I believe Company also allows this as an option?

Interesting. I'm not familiar with that company setting if so.  But I agree that with auto-commit around, it makes sense to "show what you'll be getting" in buffer.  I don't personally find auto-commit all that helpful, since I've already had to reach for an arrow key or `C-n` or whatever, hitting return doesn't seem much additional burden.  I've also disabled space from auto-committing so that makes it less useful for me anyway.

**Comment by jdtsmith** [comment]
> > BTW, should probably also call/display the doc-string for the first candidate when nothing is selected.
> 
> I am not sure. As long as you move the point to the desired candidate, I don't see a reason to show the docstring. The docstring is also noise, it is good to not show that right away.

OK this makes it clear that we use corfu differently! If I have to reach for `C-n` or an arrow key, I see that as a cache miss.  I endeavor to narrow the search and get to the right thing without the burden of hand-selecting items in a list.  It's good to see all the items to guide that selection... but interacting with it is a final resort!

**Comment by minad** [comment]
> I see that as a cache miss. I endeavor to narrow the search and get to the right thing without the burden of hand-selecting items in a list. It's good to see all the items to guide that selection... but interacting with it is a final resort!

I agree with the cache miss/slow path analogy. But if I manage to get the desired candidate at the top without using C-n (cache hit), I also don't need the docstring, since I already know what I want.

**Comment by jdtsmith** [comment]
Or, the docstring gave me the confidence it was indeed the desired thing...

**Comment by minad** [comment]
Hmm but then you are already on the slow path. I find it important that Corfu does not introduce too much noise.

**Comment by minad** [comment]
The question is if and how we can really improve on the status quo. Another alternative would be to always select the first candidate but not show it in the prompt. Then pressing down will move you to the second candidate right away and show the second candidate in the prompt, allowing auto-commit. For the first candidate auto-commit is not useful since you can press RET right away.

**Comment by minad** [comment]
Let's keep this open for now. It seems to me that what I wrote above would be a strict improvement over the status quo regarding the number of key presses. The inconsistency will move to how the overlay overwrites the prompt, but maybe this is bearable?

**Comment by jdtsmith** [comment]
> Another alternative is to always select the first candidate but not show it in the prompt. Then pressing down will move you to the second candidate right away and show the second candidate in the prompt, allowing auto-commit.

I think this could be the best solution.  So the implicitly selected 1st candidate is special-cased in terms of the prompt display.  

>The inconsistency will move to how the overlay overwrites the prompt, but maybe this is bearable?

Bearable indeed.  It would be in fact quite alarming if an implicitly selected candidate overwrote your in-buffer text without any action on your part.  So in some ways it is special cased already.  You could have tab bring that first candidate in at the prompt.  (Arrows/`C-n` etc. should probably just skip to that next candidate). 

> For the first candidate auto-commit is not useful since you can press RET right away.

Actually, on this precise topic, auto-commit for the first (implicitly selected) candidate via a commit-predicate could in fact be _very_ useful, and potentially resolve my hanging `No match` problem.  See #49 discussion.

**Comment by minad** [comment]
> Actually, on this precise topic, auto-commit for the first (implicitly selected) candidate via a commit-predicate could in fact be very useful, and potentially resolve my hanging No match problem. See #49 discussion.

I didn't take a look at your #49 proposal yet. Can this be reconciled with the behavior I described here. I don't understand how auto commit would work since the first candidate would never be explicitly selected with my proposal.

**Comment by jdtsmith** [comment]
> > Actually, on this precise topic, auto-commit for the first (implicitly selected) candidate via a commit-predicate could in fact be very useful, and potentially resolve my hanging No match problem. See #49 discussion.
> 
> I didn't take a look at your #49 proposal yet. Can this be reconciled with the behavior I described here? I don't understand how auto commit would work since the first candidate would never be explicitly selected with my proposal.

I thought your proposal was in fact to auto-select the 1st candidate, but simply not include it in the prompt overlay?  

My #49 current idea is to allow a commit predicate to commit the implicit (and now highlighted) 1st match, even when there is no "real" selection.  It's working really well in my tests.  For simple things like variable names that match exactly and come to the top, just commit them.  That one is a no-brainer because the text in buffer already matches the 1st candidate.  

In addition, with my current #49 predicate, if you type a _really good_ search string and whittle the list down to one candidate, just type on (space, close paren, etc.) and it auto-commits.  Reduces the slow path/cache miss for me substantially.   Some may not like this since the result isn't "previewed" in the buffer.  

But the key thing is it's configurable for anyone via their personal commit-predicate.  Some users will like to commit only "real" selected candidates, and always, others will like to auto-commit the implicit first selection in various cases, others will want to insist on an explicit Return to change buffer contents. Everybody happy.

**Comment by minad** [comment]
Okay, so you only want to autocommit the first candidate if it is an exact match? This should certainly be compatible with my proposal here. I am a bit biased against this idea since exact candidates are often substrings of longer candidates. Furthermore one may want to implement more intelligent auto commit mechanisms, but this reminds me a lot of mobile phone auto correction. I am very much opposed to such an auto commit functionality ;)

But it shouldn't be hard to accommodate this functionality since it is just a slight generalization/simplification of the `corfu-commit-predicate` as it is used now.

**Comment by jdtsmith** [comment]
Only when the next key is among certain "I'm moving on" keys like space, close parens, etc.  My current predicate (requires modifying corfu to run even when there is no real selection):

```elisp
(defun my/corfu-commit-predicate ()
  (if (< corfu--index 0) ; Nothing yet selected
      (cond
       ((eq corfu--total 1) ;just one candidate, commit for a few things
	(seq-intersection (this-command-keys-vector) [?: ?, ?\) ?\]]))
      
       ((and corfu--input ; exact 1st match: commit on space/etc.
	     (string-equal (string-trim-right (car corfu--input))
			   (car corfu--candidates)))
	(seq-intersection (this-command-keys-vector) [?: ?, ? ?\) ?\]])))
    (not (seq-contains-p (this-command-keys-vector) ? )))) ;let space continue
```

**Comment by jdtsmith** [comment]
>  this reminds me a lot of mobile phone auto correction. I am very much opposed to such an auto commit functionality ;)

I guess you're not a fan of "auto-commit fail" meme-genre ;)?  Agreed!

**Comment by jdtsmith** [comment]
> But it shouldn't be hard to accommodate this functionality since it is just a slight generalization/simplification of the `corfu-commit-predicate` as it is used now.

In fact it is rather simple and only requires not testing for selection before running the predicate (see #49).

**Comment by minad** [comment]
I would like to emphasize - if it is possible to accommodate these features by generalizing/simplifying existing functionality I am all for it. If it complicates things and adds more and more special conditions, we should not add the feature. There is also `corfu-quit-at-boundary` which you seem to replicate here. Of course `corfu-quit-at-boundary` is much weaker since it doesn't allow orderless. So if `corfu-commit-predicate` is a more general substitute we could even remove `corfu-quit-at-boundary`. The problem is that `corfu-commit-predicate` predicate may be backend/mode dependent due to the terminator characters, which is a bit ugly.

**Comment by jdtsmith** [comment]
Full agreement on complexity avoidance.  We could make auto-commit even more general and allow it to return 'quit, meaning:

- 'quit: exit the popup without inserting anything
- non-nil: commit (possibly implicit 1st) selection
- nil: continue on with your filtering

In fact this could entirely subsume quit no-match, quit at boundary, etc. (though no harm in leaving them). 

This nice things is that all the complexity/ugliness here is contained in the user's configuration.  Having some sensible and conservative default will generate the least surprise.  Then when users wish for additional behavior: point them to the wiki where various flavors of auto-completion are discussed.

**Comment by minad** [comment]
I experimented with my idea https://github.com/minad/corfu/issues/45#issuecomment-895457074 in branch https://github.com/minad/corfu/tree/rework-first-selection. 

Unfortunately there is one issue. TAB usually inserts the selected candidate. If the prompt is selected, `try-completion` is used instead. When the first candidate is always selected it is either not possible to TAB-complete or not possible to insert the current candidate. See:

https://github.com/minad/corfu/blob/a590f98b6ebcc8cf9e0be6abe0f43ccf92e474c9/corfu.el#L804

However this behavior has been some kind of magic trick before, which reconciled Vertico-style always insert on TAB and prefix completion. It would probably be cleaner to always insert and offer the prefix expansion as a separate command which can be bound optionally for people who prefer that style of completion.

**Comment by minad** [comment]
There is one important reason though why I went with this magic combination of `try-completion` and insertion if an actual candidate is selected: When Corfu completion is triggered via TAB (if you don't use auto), then TAB initiates a prefix completion. So it is only natural to continue supporting prefix completion instead of going with the eager insertion of the current candidate. This behavior plays mostly a role when completing paths in shell/eshell. However the current behavior is also not perfect - it happens often that one presses TAB and nothing happens since no prefix expansion is possible.

**Comment by minad** [comment]
Also the new behavior with preselection of the first candidate leads to bad behavior if one accidentally presses TAB TAB, since then the first candidate is inserted directly.

**Comment by jdtsmith** [comment]
> When Corfu completion is triggered via TAB (if you don't use auto), then TAB initiates a prefix completion.

I don't see this.  For me M-Tab=indent-for-tab-command does full orderless completion (though not crossing word boundaries).

**Comment by minad** [comment]
This happens only if you set `tab-always-indent=complete`.

**Comment by jdtsmith** [comment]
>  if one accidentally presses TAB TAB, since then the first candidate is inserted directly.

In the corfu keymap TAB cycles through candidates.  Why would that not be the case if the 1st is auto-selected?

**Comment by minad** [comment]
It either cycles, inserts or expands.

https://github.com/minad/corfu/blob/1c1a176bf601657e1cc8be48fa833d5ce6b14b4c/corfu.el#L802-L804

**Comment by jdtsmith** [comment]
> This happens only if you set `tab-always-indent=complete`.

Tried that.  Same behavior (replacing M-Tab with Tab).

**Comment by minad** [comment]
This is because of your configuration:

~~~ elisp
  ;; Optionally use TAB for cycling, default is `corfu-complete'.
  ;; :bind (:map corfu-map
  ;;        ("TAB" . corfu-next)
  ;;        ([tab] . corfu-next)
  ;;        ("S-TAB" . corfu-previous)
  ;;        ([backtab] . corfu-previous))
~~~

**Comment by minad** [comment]
To bring the discussion back on track - the question is if there is a possibility to fit TAB-prefix completion (`try-completion` expansion) into the framework. By preselecting of the first candidate this is not possible, since `corfu-complete` is overloaded for insertion, cycling and expansion. Of course this only happens if you don't bind TAB to `corfu-next`.

**Comment by jdtsmith** [comment]
True.  Seems like a sensible default config to me.  One oddity in the rework branch: Ret doesn't select the candidate (1st or otherwise).

**Comment by minad** [comment]
> True. corfu-next seems like a sensible default config to me. 

No, it is actually not such a good default when you want to complete paths in shells.

> One oddity in the rework branch: Ret doesn't select the candidate (1st or otherwise).

It does. Did you compile the package properly? The corfu-map changed.

> try-completion doesn't respect the completion style?

I am using `completion-try-completion` of course.

**Comment by jdtsmith** [comment]
> > True. corfu-next seems like a sensible default config to me.
> 
> No, it is actually not such a good default when you want to complete paths in shells.

Aha.  What is the objective?  To be able to Tab-complete directories as in the shell and leave corfu out of it, even if there are multiple matches for a given directory?  company and its file backend finesse this by

1. pre-filling single candidate results in the buffer with underlining.
2. "auto-completing" such single candidates upon Tab.

With a general command predicate you could achieve 2 easily.  Tab auto accepts any single candidate.  Otherwise it will "cycle" using `corfu-next`. 

> > One oddity in the rework branch: Ret doesn't select the candidate (1st or otherwise).
> 
> It does. Did you compile the package properly? The corfu-map changed.

Working now.  Had to disable/reenable corfu to pick up the new map.

**Comment by minad** [comment]
>  What is the objective? 

It is the same as in Vertico, where you can insert the current candidate with TAB. This is used in `find-file` for example. The goal is not to keep Corfu out of the completion, but to continue completion even after inserting a candidate. Since there is the concept of prompt selection currently, I can overload this and use TAB for `completion-try-completion`. This is the main feature which would be lost if I always preselect a candidate. Of course I could bind a `completion-try-completion` command to a separate key. This is in fact what I recommend for Vertico (https://github.com/minad/vertico#completion-styles-and-tab-completion), see the `minibuffer-complete` binding.

**Comment by jdtsmith** [comment]
But unlike vertico, corfu appears to place the current selection in the buffer always.  So from the user's perspective, it's just a matter of how you "accept"/"complete" a given candidate.  I agree that tab is indeed quite convenient for that.  

Regarding my point 2 above, my auto-commit predicate has no access to Tab, since it invokes one of the omitted commands by default (corfu-next).  I could always remove that from the list.  I'll think about this some and if I have any deep thoughts report back.  

My summary so far:

1. An implicit `Ret`-able 1st match without any highlight may be confusing to the user. 
2. But there also are real advantages to maintaining a "nothing is selected" state.
3. A UI-only "soft" highlight isn't a great solution. 
3. `corfu-commit-predicate` can be made more general if it is allowed to quit, commit, or continue filtering (`corfu-commit-quit-continue-predicate`?), and if it is called even without any active selection (other than the implicit 1st candidate selection).  This may lead to other simplifications. 
4. Tab is being overworked for initiating completion, cycling candidates, and inserting file-path candidates.

**Comment by minad** [comment]
I thought about this a bit more and it seems there is just no way to get rid of the concept of "prompt selection". The prompt is selected when no candidate is highlighted. When I complete directories, the prompt is always selected such that you can finish with the current directory with RET. If I would always preselect the current candidate, then I could not select the current directory. If I remember correctly Company gets around this restriction by always showing the entire path for completion.

The current behavior seems to be quite reasonable to me and it seems not possible to improve on this approach. 
Also the idea of visual soft selection is not useful in the context of path completion. There are these two issues:

1. When scrolling down to a candidate below the first, one more key press is required than in Company. On the other hand the currently selected candidate is shown in the prompt in Corfu which is a nice feature.
2. The first candidate gets a bit of special treatment (It is softly selected as you write). If you press RET the candidate is inserted, and if you press TAB without progress, the candidate is completed (just added this behavior, https://github.com/minad/corfu/commit/68ea28e987f4079f45916800f1a7b9530600a198).

**Comment by jdtsmith** [comment]
> I thought about this a bit more and it seems there is just no way to get rid of the concept of "prompt selection". The prompt is selected when no candidate is highlighted. 

I see that state as "nothing is selected".  There is no indication that the prompt is selected, either via face highlight or (obviously) with overlaid text at the prompt/in the buffer. 

> When I complete directories, the prompt is always selected such that you can finish with the current directory by pressing RET. If I would always preselect the current candidate, then one could not finish completion with the current directory. If I remember correctly Company gets around this restriction by always showing the entire path for completion.

I do see this point and agree with it.  My soft highlighting idea would preserve all this but give some indication of what would happen if you hit Ret.  But it has drawbacks and unexpected behavior as you note. 

Regarding what Tab should do: once I disabled the `corfu-next` etc. bindings, `corfu-complete` already seems to do close to what one might want.  I.e. with only one _directory_  candidate, Tab auto-commits it.  I would want that behavior symmetrically everywhere.  I.e., I wish Tab when there is only one candidate in any situation would always auto-commit that candidate (otherwise `corfu-next`'ing).  The could probably be done in a commit predicate if we allow these to "commit" the implicit selection when at prompt. Right now they are not even run unless there is an explicit selection.

> The current behavior seems to be quite reasonable to me and it seems not possible to improve on this approach.

I agree.  Users (me included) will just have to get used to it :).   The only idea which might help them is to somehow indicate visually that the prompt itself is "active".  But I can't think of a way to do this that isn't ugly. 

> 1. When scrolling down to a candidate below the first, one more key press is required than in Company. On the other hand the currently selected candidate is shown in the prompt in Corfu.

The advanced filtering capability like orderless more than makes up for this. 

> 2. The first candidate gets a bit of special treatment (It is softly selected as you say). If you press RET the candidate is inserted, and if you press TAB without progress, the candidate is completed (just added this behavior, [68ea28e](https://github.com/minad/corfu/commit/68ea28e987f4079f45916800f1a7b9530600a198)).

Will give a look.

**Comment by minad** [comment]
> I see that state as "nothing is selected". There is no indication that the prompt is selected, either via face highlight or (obviously) with overlaid text at the prompt/in the buffer.

I don't perceive it as such. The popup without highlight is not taking focus, since it doesn't have a highlight. This means the focus is still there were it has been before - at the point!

> I do see this point and agree with it. My soft highlighting idea would preserve all this but give some indication of what would happen if you hit Ret. But it has drawbacks and unexpected behavior as you note.

As I see it the highlight does not really fit the current design. You should try shell path completion. And since the soft highlight adds only eye candy without further behavioral improvements, I already came to the conclusion that this should not be added.

> The only idea which might help them is to somehow indicate visually that the prompt itself is "active". But I can't think of a way to do this that isn't ugly.

I don't see a reason to do this. I also think it would mainly add visual noise.

**Comment by jdtsmith** [comment]
I tried the new _Tab-again_ update, and found it to be useful in some cases, but also leading to some inconsistencies between partial completion and orderless that are distracting.  

With the default Tab binding to `corfu-complete`, I observe Tab behavior as follows (annotated with my perceptions as positive (+) and negative (-)):

- In filepath (partial-completion):
  -  If you have entered additional "search text" beyond what prompted corfu to pop up:
      - (+) If only one directory remaining, a single Tab commits it.
      - (+) If several directories or files remain, Tab completes any common prefix, and another Tab commits the first candidate. 
      - (-) When completing a final file at the _end_ of a file-path, even when there is just one candidate, Tab inserts that single candidate, but never commits it.  Subsequent Tabs do not either. This differs from subdirectory behavior. 
  - If you have _not_ entered any new search text:
    - (-) Any number of Tabs does nothing, _even if there is only one candidate_. 
- Elsewhere (orderless):
  - (+) with only one candidate, Tab commits it.
  - (-) with multiple candidates showing, Tab does nothing, 
  - (+) Double-Tab inserts the first candidate, and then commits it, unless it is a common prefix. 

What I'd prefer:

- One Tab: If there is only one candidate in any completion context (filepath subdirectory + file, or otherwise), Tab commits it in all cases, whether or not additional search text has been entered.
- More Tabs:
  - (Option 1, my preference) With more than one candidate, Tab cycles through them (with `corfu-next`). 
  - (Or, Option 2) Double-tab inserts the 1st candidate, committing it unless it's a common prefix, as it does now.

I should pause here to thank you for all the incredible work you are doing with corfu/vertico/consult/marginalia/etc.  I think corfu already serves as a much lighter-weight and easier to configure replacement for the traditional completion in region and company both.  All of my thoughts here represent tweaks on already quite good functionality.  My favorite feature is being able to use the same completion styles for minibuffer completion and in-buffer completion.  Increased power and reduced context switching overhead. Thanks!

**Comment by minad** [comment]
Thanks for these proposals. Just to let you know, I will let Corfu rest for a while and look into it again at a later time - concentrating on other things right now. It would also be good to hear feedback by other users.

**Comment by salutis** [comment]
Chiming in from #56. This issue is a big problem for me, as Corfu intrusively changes the text I type to nonsense. I find it helpful to see the available completions at all times, but I want Corfu to _never_ change the text I typed, except when I give it a command to do so, such as `C-n` and `RET`.

**Comment by jdtsmith** [comment]
I'm not sure I follow this. What nonsense? Do you mean corfu "implicitly" accepts the first candidate upon hitting Return?  Unless you hit return, no candidate info is filled in the buffer.

**Comment by salutis** [comment]
@jdtsmith I apologize for unclear communication. When I hit `RET`, I expect Emacs to insert a newline into the buffer. To put it differently, Corfu should either (1) highlight the first candidate to communicate its pre-selection to the user or (2) avoid binding the first candidate to `RET` by default. Personally, I would go for (1), as seen virtually everywhere else, including in all mainstream IDEs.

**Comment by minad** [comment]
@salutis (1) is exactly what @jdtsmith proposed. See https://github.com/minad/corfu/issues/45#issuecomment-895876009 for the issue with pre-selection. For (2), to get rid of the magic it probably suffices to unbind RET or to rebind it to `newline`.

**Comment by salutis** [comment]
Daniel Mendler ***@***.***> writes:

> (1) is exactly what @jdtsmith proposed. See 
> https://github.com/minad/corfu/issues/45#issuecomment-895876009 
> for the issue with pre-selection.

I need to try to read it again, for I got lost in the argument.

> For (2), to get rid of the magic it probably suffices to unbind 
> RET or to rebind it to `newline`.

I did `(define-key corfu-map (kbd "RET") nil)` and the magic 
disappeared, but I now cannot do `C-n RET` either. 

-- 
Programming reliably --- must be an activity of an undeniably 
mathematical nature [...] You see, mathematics is about thinking, 
and doing mathematics is always trying to think as well as 
possible. -- Edsger W. Dijkstra (1981)  Rudolf Adamkovič 
***@***.***> Studenohorská 25 84103 Bratislava Slovakia 
[he/him]

**Comment by jdtsmith** [comment]
I've personally gotten used to "Ret selects the first thing" which is borrowed from vertico.  In vertico it of course makes more sense, and I note vertico _does_ highlight the 1st match "by default".   But that is because it starts with a _real selection_, equivalent to having `C-n`'ed once with corfu.  And in vertico you can up arrow to get back "to the prompt", which is useful for filling in something else that doesn't even exist among the candidates (like a new filename).  The confusing thing about corfu is it starts in this "prompt selected" state.   It does so because it fills in the candidate in the buffer/at the prompt itself.  Vertico does not do that: the prompt is always just for filter text.  

Maybe a corfu mode that replicates vertico in this manner would satisfy people confused by "Ret implicitly selects 1st prompt".

**Comment by minad** [comment]
@salutis You can use a command like this, which either commits or inserts a newline:

~~~ elisp
(defun corfu-insert ()
  (interactive)
  (if (>= corfu--index 0)
      (corfu--insert 'finished)
    (corfu-quit)
    (newline)))
~~~

**Comment by minad** [comment]
I am experimenting with an alternative approach in https://github.com/minad/corfu/pull/78. There is a flag `corfu-preview-current` which can be used to disable the current candidate overlay. If the overlay is disabled the first candidate can be preselected if the prompt itself is not a valid completion.

**Comment by salutis** [comment]
> @salutis You can use a command like this, which either commits or inserts a newline: [...]

Thank you, I will give it a shot!

**Comment by minad** [comment]
Addressed in #88. @salutis @jdtsmith I would appreciate some feedback.

**Comment by salutis** [comment]
> Addressed in #88. @salutis @jdtsmith I would appreciate some feedback.

Wow, I did not even notice the change and thus did to reply to this
issue.  Why?  My confusion with the selection went /poof/, but my
brain did not tell me, because everything now works "as expected".
Smooth.  Thank you, @minad!
